File: internal/report/attributes.go
package report

import (
	"fmt"
	"sort"
	"strings"
)

// attributeLabels, connector'ların Context anahtarlarını okunabilir Türkçe
// etiketlere çevirir. Buradaki sıra aynı zamanda GÖSTERİM ÖNCELİĞİDİR:
// üstteki anahtarlar bir kişi hakkında daha çok şey söyler.
var attributeLabels = []struct {
	Key   string
	Label string
}{
	// Kimlik — bir araştırmacının önce görmek istediği şeyler
	{"name", "Ad"},
	{"full_name", "Tam ad"},
	{"display_name", "Görünen ad"},
	{"profile_username", "Profil adı"},
	{"bio", "Biyografi"},
	{"location", "Konum"},
	{"company", "Şirket"},
	{"email", "E-posta"},
	{"blog", "Web sitesi"},
	{"twitter_username", "X/Twitter"},
	{"avatar", "Avatar"},

	// Hesap sinyalleri — yaş ve büyüklük güvenilirlik göstergesidir
	{"created", "Hesap açılışı"},
	{"followers", "Takipçi"},
	{"following", "Takip edilen"},
	{"repos", "Repo sayısı"},
	{"platform", "Platform"},
	{"verification", "Doğrulama"},
	{"match", "Eşleşme"},
	{"independent_sources", "Bağımsız kaynak grubu"},
	{"confidence_basis", "Güven gerekçesi"},

	// Teknik varlıklar
	{"repo", "Repo"},
	{"language", "Dil"},
	{"org", "Organizasyon"},
	{"isp", "ISP"},
	{"os", "İşletim sistemi"},
	{"city", "Şehir"},
	{"country", "Ülke"},
	{"ports", "Açık portlar"},
	{"tags", "Etiketler"},
	{"vuln_count", "Zafiyet sayısı"},
	{"product", "Ürün"},
	{"version", "Sürüm"},
	{"ip", "IP"},
}

// displayNoiseKeys, GÖSTERİMDEN çıkarılan ama veride kalan anahtarlardır.
//
// Veriden silmiyoruz — API/JSON çıktısında dursunlar. Sadece insan okuyan
// tabloda yer kaplamasınlar:
//   - "found": raporlanmış her bulgu için zaten true, bilgi taşımıyor
//   - "username": aranan hedefin kendisi, her satırda tekrarlanıyor
//   - "source": zaten Entity.Sources sütununda gösteriliyor
//
// "source" resolution katmanında da eleniyor; burada tekrar elenmesi
// bilinçlidir — gösterim, yukarıdaki katmanın filtrelemesine güvenmemeli.
var displayNoiseKeys = map[string]bool{
	"found":    true,
	"username": true,
	"source":   true,
}

// identityKeys, "kişi özeti" bloğunda toplanacak anahtarlardır.
var identityKeys = map[string]bool{
	"name": true, "full_name": true, "display_name": true,
	"bio": true, "location": true, "company": true,
	"email": true, "blog": true, "twitter_username": true,
	"created": true, "followers": true,
	"profile_username": true,
}

// Attribute, gösterime hazır tek bir nitelik satırıdır.
type Attribute struct {
	Key   string
	Label string
	Value string

	// Source, bu iddianın HANGİ platformdan geldiğidir.
	//
	// Kimlik özetinde zorunludur: farklı platformlardaki aynı kullanıcı adı
	// aynı kişi DEĞİLDİR. Kaynak gösterilmezse birbiriyle ilgisiz insanların
	// bilgileri tek bir kimlik profiline karışır (bkz. IdentitySummary).
	Source string
}

// formatValue, JSON'dan gelen any değerini okunabilir tek satıra çevirir.
// json.Unmarshal tüm sayıları float64 yaptığı için tam sayılar ".00" olmadan
// basılır — "18.00 takipçi" saçma görünüyordu.
func formatValue(v any) string {
	switch t := v.(type) {
	case nil:
		return ""
	case string:
		return strings.TrimSpace(t)
	case float64:
		if t == float64(int64(t)) {
			return fmt.Sprintf("%d", int64(t))
		}
		return fmt.Sprintf("%g", t)
	case bool:
		if t {
			return "evet"
		}
		return "hayır"
	case []any:
		parts := make([]string, 0, len(t))
		for _, item := range t {
			if s := formatValue(item); s != "" {
				parts = append(parts, s)
			}
		}
		return strings.Join(parts, ", ")
	default:
		return fmt.Sprintf("%v", t)
	}
}

// OrderedAttributes, bir entity'nin niteliklerini gösterim sırasına dizer.
// attributeLabels'ta tanımlı olanlar önce ve o sırayla; tanımsız anahtarlar
// sona alfabetik olarak eklenir (yeni connector'lar sessizce kaybolmasın).
func OrderedAttributes(attrs map[string]any) []Attribute {
	if len(attrs) == 0 {
		return nil
	}

	out := make([]Attribute, 0, len(attrs))
	seen := make(map[string]bool, len(attrs))

	for _, def := range attributeLabels {
		v, ok := attrs[def.Key]
		if !ok || displayNoiseKeys[def.Key] {
			continue
		}
		seen[def.Key] = true
		if s := formatValue(v); s != "" {
			out = append(out, Attribute{Key: def.Key, Label: def.Label, Value: s})
		}
	}

	rest := make([]string, 0)
	for k := range attrs {
		if !seen[k] && !displayNoiseKeys[k] {
			rest = append(rest, k)
		}
	}
	sort.Strings(rest)
	for _, k := range rest {
		if s := formatValue(attrs[k]); s != "" {
			out = append(out, Attribute{Key: k, Label: k, Value: s})
		}
	}

	return out
}

// IdentitySummary, kimlik iddialarını KAYNAĞIYLA BİRLİKTE listeler.
//
// Bu fonksiyon önce tüm platformlardan gelen değerleri tek bir profile
// birleştiriyordu. Bu YANLIŞTI ve canlı testte zararı görüldü: t.me/testuser
// gerçekten var ama BAŞKA birine ait; "Deniz Kaya" adı, araştırılan kişinin
// kimlik özetine sanki onunmuş gibi yazıldı. Aynı şekilde
// instagram.com/testuser bambaşka bir hesap — hedefin gerçek hesabı
// _testuser_ idi.
//
// Temel OSINT kuralı: FARKLI PLATFORMLARDAKİ AYNI KULLANICI ADI AYNI KİŞİ
// DEĞİLDİR. Bu yüzden artık birleştirme yapılmıyor; her iddia kendi kaynağıyla
// ayrı satır olarak gösteriliyor ve çelişkiler araştırmacıya görünür kalıyor.
//
// Aynı değer birden çok platformda geçiyorsa tek satırda toplanır — bu
// gerçek bir doğrulama sinyalidir.
func IdentitySummary(entities []EntityData) []Attribute {
	// key → değer → o değeri bildiren kaynaklar (sıra korunur)
	claims := make(map[string]map[string][]string)
	order := make(map[string][]string)

	for _, e := range entities {
		src := entitySourceLabel(e)
		for k, v := range e.Attributes {
			if !identityKeys[k] {
				continue
			}
			val := formatValue(v)
			if val == "" {
				continue
			}
			if claims[k] == nil {
				claims[k] = make(map[string][]string)
			}
			if _, seen := claims[k][val]; !seen {
				order[k] = append(order[k], val)
			}
			claims[k][val] = appendUniqueStr(claims[k][val], src)
		}
	}

	var out []Attribute
	for _, def := range attributeLabels {
		values, ok := claims[def.Key]
		if !ok {
			continue
		}
		for _, val := range order[def.Key] {
			out = append(out, Attribute{
				Key:    def.Key,
				Label:  def.Label,
				Value:  val,
				Source: strings.Join(values[val], ", "),
			})
		}
	}
	return out
}

// entitySourceLabel, bir iddianın hangi platformdan geldiğini adlandırır.
func entitySourceLabel(e EntityData) string {
	if p, ok := e.Attributes["platform"]; ok {
		if s := formatValue(p); s != "" {
			return s
		}
	}
	if len(e.Sources) > 0 {
		return e.Sources[0]
	}
	return "bilinmiyor"
}

func appendUniqueStr(list []string, v string) []string {
	for _, x := range list {
		if x == v {
			return list
		}
	}
	return append(list, v)
}

// SortByInformation, entity'leri BİLGİ YOĞUNLUĞUNA göre sıralar: gerçek
// nitelik taşıyanlar üstte, çıplak URL'ler altta.
//
// Önceden sıralama map iterasyon sırasına bağlıydı, yani rastgeleydi. Pratikte
// asıl değerli kayıt (ad, konum, hesap yaşı içeren profil) 25 satırlık bir
// URL listesinin ortasında kayboluyordu.
func SortByInformation(entities []EntityData) {
	sort.SliceStable(entities, func(i, j int) bool {
		ni := len(OrderedAttributes(entities[i].Attributes))
		nj := len(OrderedAttributes(entities[j].Attributes))
		if ni != nj {
			return ni > nj
		}
		if entities[i].Type != entities[j].Type {
			return entities[i].Type < entities[j].Type
		}
		return entities[i].Value < entities[j].Value
	})
}

// CompactAttributes, tablo hücresine sığacak kısa bir özet üretir.
func CompactAttributes(attrs map[string]any, max int) string {
	ordered := OrderedAttributes(attrs)
	if len(ordered) == 0 {
		return ""
	}
	if max > 0 && len(ordered) > max {
		ordered = ordered[:max]
	}

	parts := make([]string, 0, len(ordered))
	for _, a := range ordered {
		v := a.Value
		if len(v) > 40 {
			v = v[:37] + "..."
		}
		parts = append(parts, a.Label+": "+v)
	}
	return strings.Join(parts, " · ")
}

File: ui/src/lib/attributes.js
// Nitelik gösterim kuralları.
//
// Go tarafındaki internal/report/attributes.go ile aynı mantığı uygular:
// sıra = gösterim önceliği, gürültü anahtarları gizlenir, boş değerler elenir.

const LABELS = [
  // Kimlik — araştırmacının önce görmek istedikleri
  ['name', 'Ad'],
  ['full_name', 'Tam ad'],
  ['display_name', 'Görünen ad'],
  ['profile_username', 'Profil adı'],
  ['bio', 'Biyografi'],
  ['location', 'Konum'],
  ['company', 'Şirket'],
  ['email', 'E-posta'],
  ['blog', 'Web sitesi'],
  ['twitter_username', 'X/Twitter'],

  // Hesap sinyalleri
  ['created', 'Hesap açılışı'],
  ['followers', 'Takipçi'],
  ['following', 'Takip edilen'],
  ['repos', 'Repo sayısı'],
  ['platform', 'Platform'],

  // Teknik
  ['repo', 'Repo'],
  ['language', 'Dil'],
  ['org', 'Organizasyon'],
  ['isp', 'ISP'],
  ['city', 'Şehir'],
  ['country', 'Ülke'],
  ['ports', 'Açık portlar'],
  ['tags', 'Etiketler'],
  ['vuln_count', 'Zafiyet sayısı'],
  ['ip', 'IP'],
];

// Kart yüzeyinde gösterilmeyenler: ya ayrı bir bileşende gösteriliyorlar
// (avatar, güven gerekçesi, rozet) ya da tekrar eden gürültüler.
const HIDDEN = new Set([
  'found', 'username', 'source', 'avatar', 'site',
  'confidence_basis', 'confidence_groups', 'confidence_logodds',
  'independent_sources', 'verification', 'match', 'extracted',
]);

const IDENTITY_KEYS = new Set([
  'name', 'full_name', 'display_name', 'profile_username',
  'bio', 'location', 'company', 'email', 'blog',
  'twitter_username', 'created', 'followers',
]);

export function formatValue(v) {
  if (v === null || v === undefined) return '';
  if (typeof v === 'boolean') return v ? 'evet' : 'hayır';
  if (Array.isArray(v)) return v.map(formatValue).filter(Boolean).join(', ');
  if (typeof v === 'number') return Number.isInteger(v) ? String(v) : String(v);
  return String(v).trim();
}

/** Bir varlığın niteliklerini gösterim sırasına dizer. */
export function orderedAttributes(attrs = {}) {
  const out = [];
  const seen = new Set();

  for (const [key, label] of LABELS) {
    if (HIDDEN.has(key) || !(key in attrs)) continue;
    seen.add(key);
    const value = formatValue(attrs[key]);
    if (value) out.push({ key, label, value });
  }

  // Tanımsız anahtarlar sessizce kaybolmasın.
  for (const key of Object.keys(attrs).sort()) {
    if (seen.has(key) || HIDDEN.has(key)) continue;
    const value = formatValue(attrs[key]);
    if (value) out.push({ key, label: key, value });
  }

  return out;
}

/**
 * Kimlik iddialarını KAYNAĞIYLA listeler.
 *
 * Birleştirme YAPILMAZ. Farklı platformlardaki aynı kullanıcı adı aynı kişi
 * değildir: canlı testte t.me/testuser gerçekten vardı ama başka birine
 * aitti, ve "Deniz Kaya" adı araştırılan kişinin kimliğiymiş gibi
 * gösterilmişti. Her iddia kendi kaynağıyla ayrı satır olarak durur.
 */
export function identityClaims(entities = []) {
  const claims = new Map(); // key -> Map(value -> Set(source))

  for (const e of entities) {
    const attrs = e.attributes || {};
    const source = attrs.platform || (e.sources && e.sources[0]) || 'bilinmiyor';

    for (const [key, value] of Object.entries(attrs)) {
      if (!IDENTITY_KEYS.has(key)) continue;
      const text = formatValue(value);
      if (!text) continue;

      if (!claims.has(key)) claims.set(key, new Map());
      const byValue = claims.get(key);
      if (!byValue.has(text)) byValue.set(text, new Set());
      byValue.get(text).add(source);
    }
  }

  const out = [];
  for (const [key, label] of LABELS) {
    const byValue = claims.get(key);
    if (!byValue) continue;
    for (const [value, sources] of byValue) {
      out.push({ key, label, value, sources: [...sources] });
    }
  }
  return out;
}

/** Güven puanı → renk/etiket bandı. */
export function confidenceBand(score) {
  if (score >= 80) return { level: 'high', label: 'Yüksek' };
  if (score >= 60) return { level: 'medium', label: 'Orta' };
  if (score >= 35) return { level: 'low', label: 'Düşük' };
  return { level: 'minimal', label: 'Çok düşük' };
}

/** Varlığın taşıdığı uyarı rozetleri. */
export function badges(entity) {
  const attrs = entity.attributes || {};
  const out = [];

  if (attrs.match) out.push({ kind: 'variant', text: 'varyant eşleşme' });
  if (attrs.verification) out.push({ kind: 'suspect', text: attrs.verification });
  if (attrs.extracted) out.push({ kind: 'derived', text: 'biyografiden çıkarıldı' });

  return out;
}

/** Bilgi yoğunluğuna göre sıralama — dolu kayıtlar üstte. */
export function sortByInformation(entities = []) {
  return [...entities].sort((a, b) => {
    const na = orderedAttributes(a.attributes).length;
    const nb = orderedAttributes(b.attributes).length;
    if (na !== nb) return nb - na;
    if (b.confidence !== a.confidence) return b.confidence - a.confidence;
    return (a.primary_value || '').localeCompare(b.primary_value || '');
  });
}
